src/flakeref/forge.rs

```diff
@@ -0,0 +1,201 @@
+use std::{fmt::Display, path::Path};
+
+use nom::{
+    branch::alt,
+    bytes::complete::{tag, take_until},
+    combinator::{map, opt, rest},
+    multi::many_m_n,
+    IResult,
+};
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    error::{NixUriError, NixUriResult},
+    parser::parse_url_type,
+};
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub enum GitForgePlatform {
+    GitHub,
+    GitLab,
+    SourceHut,
+}
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub struct GitForge {
+    pub platform: GitForgePlatform,
+    pub owner: String,
+    pub repo: String,
+    pub ref_or_rev: Option<String>,
+}
+
+impl GitForgePlatform {
+    fn parse_hub(input: &str) -> IResult<&str, Self> {
+        map(tag("github"), |_| Self::GitHub)(input)
```

Comment:
Does it make sense to use the Display impl from `Self::*` here or is that too convoluted?

Reply:
At some point, using `strum` would make sense. That, or some other "clean code" idiom. Might be worth rolling up into this PR. I suspect it will be a bit out of scope for this one.

Reply:
#159